### ISSUE TITLE
jmap not happy on alpine

### ISSUE THREAD

**Issue by devinrsmith** [docker-library/openjdk#76]
We ran into an issue today where we needed to take a heap dump (jmap). But it wasn't working.

Image was `java:openjdk-8u92-jdk-alpine`.

```
$ docker exec server1_v3_1 ps aux
PID   USER     TIME   COMMAND
    1 root       1:54 java -server -verbose:gc ...
  121 root       0:00 ps aux

$ docker exec server1_v3_1 jmap -dump:live,format=b,file=/data/dump.map 1
1: Unable to get pid of LinuxThreads manager thread
```

Here's the source of the error: http://code.metager.de/source/xref/openjdk/jdk8/jdk/src/solaris/native/sun/tools/attach/LinuxVirtualMachine.c#255 

We suspect it might be because alpine is using musl instead of glibc. 

jmap works in `java:openjdk-8u91-jdk`.

**Comment by niloc132** [comment]
Other tools that don't work: `jstack`, `jinfo`, `jhat`.

**Comment by paranoid945** [comment]
Does any openjdk engineer watching this issue?

**Comment by yosifkit** [comment]
This seems related: https://github.com/gliderlabs/docker-alpine/issues/11.

**Comment by michaelzg** [comment]
Tried what @yosifkit linked. 

Installed `glibc` as suggested by sgerrand in https://github.com/sgerrand/alpine-pkg-glibc/issues/1 and still no luck:

```
/opt # apk info | grep glibc
glibc
glibc-bin

/opt # java -version
openjdk version "1.8.0_111-internal"
OpenJDK Runtime Environment (build 1.8.0_111-internal-alpine-r0-b14)
OpenJDK 64-Bit Server VM (build 25.111-b14, mixed mode)

/opt # ps aux
PID   USER     TIME   COMMAND
    1 root       8:38 java -Xms2g -Xmx2g -XX:+UseG1GC -Dconfig.file=application.conf -jar wss.jar
  303 root       0:00 /bin/sh
  421 root       0:00 ps aux

/opt # jmap -dump:format=b,file=heap.bin 1
1: Unable to get pid of LinuxThreads manager thread
```

**Comment by tianon** [comment]
Indeed -- I don't think there's much we can do about this. :disappointed:

**Comment by dkolog** [comment]
I noticed that too. This happens because your java process PID is 1. 
I use the same setup (alpine + glibc + jmap) and it works well if the PID is not 1 (when bash scripts forks the java process)

**Comment by yosifkit** [comment]
If you keep bash as PID 1, you'd no longer get signals from `docker stop` and `docker kill` (without having to write traps).  With docker 1.13 you can use `--init` to have docker put in [tini](https://github.com/krallin/tini) as PID 1; it'll forward signals and reap zombies.

**Comment by dkolog** [comment]
@yosifkit you are absolutely right.
@devinrsmith  I was able to get heap/thread dump from PID 1 using Oracle's JDK.

**Comment by elyscape** [comment]
This may be a result of the debug symbols not being installed in the image. [See comment 9 on this issue](https://bugs.launchpad.net/ubuntu/+source/openjdk-7/+bug/1548434) or [comment 15 on this issue](https://bugzilla.redhat.com/show_bug.cgi?id=1010786#c15).

**Comment by sgerrand** [comment]
> Installed glibc as suggested by @sgerrand 

For those arriving late to the party, like me, this issue appears to be unrelated to the underlying libc implementation.

**Comment by tianon** [comment]
Closing, since this appears to be solved by either using Docker's `--init` flag, or adding something like `tini` / `dumb-init` to images which suffer from this problem. :+1:

**Comment by jaychang9** [comment]
So how can I use jmap ?

**Comment by highel** [comment]
> So how can I use jmap ?

You can change docker file and instead of running java as `java blahlahvlha` run it as `sh -c java blahlahvlha && 1` notice `&& 1` to prevent hs from exiting,

**Comment by ScottOaks** [comment]
I disagree that this is unrelated to the libc version in alpine.

A simple docker test with FROM openjdk:8 runs the java process as PID 1, and jcmd (and similar tools) still work. GIven this dockerfile
FROM openjdk:8
COPY . /tmp/hello
WORKDIR /tmp/hello
RUN javac Hello.java
CMD ["java", "Hello"]

building and running (assuming the Hello program is long-running), then I can do this:
$ docker exec mad_mahavira jcmd 1 info
...prints expected output....

Change the dockerfile to have FROM openjdk:8-alpine and only then do I have the issue of not being able to attach to PID 1.

I'm not saying this is definitely because of the use of musl instead of glibc; it could be some way that alpine linux sets up the /proc system that is different than other linux containers. But the issue does seem to lie with Alpine.

**Comment by yashbhutwala** [comment]
@here ping!  Still facing this issue, can we please reopen this and triage the underlying issue?

**Comment by casimirrex** [comment]
hi you need to add openjdk8.xx.alpine . we are getting the same problem.

**Comment by yosifkit** [comment]
> Ping! Still facing this issue, can we please reopen this and triage the underlying issue?

No, OpenJDK 8 on Alpine are no longer maintained here since Alpine/musl in not officially supported by the OpenJDK project. See also https://github.com/docker-library/openjdk/issues/372#issuecomment-560900332.

**Comment by jhmartin** [comment]
As a google breadcrumb; https://github.com/apangin/jattach is able to generate heap dumps from within an Alpine / OpenJDK8 based container.

**Comment by ccgcool** [comment]
I too faced this issue with openjdk:8-jdk-alpine when PID was 1. Then I changed base image to maven:3.8.2-jdk-8 . It worked **even with** PID 1. So its likely not a PID being 1 issue. note: The image which solved the issue is an jdk8 image by maven, not an opened based image. For opened you may have to install the opened-devel  or opened-dbg kinda stuff because openjdk apparently does not come with debug facilities like jmap stack etc

**Comment by cliveverghese** [comment]
This issue is fixed by patch [icedtea-issue13032.patch](https://git.alpinelinux.org/aports/tree/community/openjdk8/icedtea-issue13032.patch) and on [amazoncorretto:8-alpine](https://hub.docker.com/_/amazoncorretto) as well.

**Comment by kelthuzadx** [comment]
https://github.com/kelthuzadx/jdump works for apline

**Comment by robberphex** [comment]
TL; DR:
```
pid=1 ;\
touch /proc/${pid}/cwd/.attach_pid${pid} && \
  kill -SIGQUIT ${pid} && \
  sleep 2 &&
  ls /proc/${pid}/root/tmp/.java_pid${pid}
# Attach Listener started, then you can attach this jvm normally.
```

The detail analysis is on [the post(Chinese)](https://robberphex.com/attach-jvm-in-container-at-arthas/).

**Comment by bookc-man** [comment]
run cmd add `--init` is ok

`docker run  --init `

**Comment by cyao2q** [comment]
> run cmd add `--init` is ok
> 
> `docker run --init `

Dockerfile中如何编写？

**Comment by bookc-man** [comment]
> > run cmd add `--init` is ok
> > `docker run --init `
> 
> Dockerfile中如何编写？

正常写就行了
例如
`
FROM openjdk-8-alpine:latest

COPY ../target/management.jar /opt/project/management.jar

WORKDIR /opt/project

CMD java -jar management.jar
`

**Comment by cyao2q** [comment]
> 

Dockerfile中如何使用 --init 参数  还是 只能docker run的时候使用 

目前可以使用 tini 插件解决 就是感觉 不想安装

```
FROM openjdk:8-jdk-alpine3.9

COPY *.jar /app.jar

RUN apk --no-cache add tini

CMD java -Dfile.encoding=UTF-8 -Duser.timezone=PRC -Dserver.port=8080 -jar /app.jar

ENTRYPOINT ["tini"]

```

**Comment by bookc-man** [comment]
> > 
> 
> Dockerfile中如何使用 --init 参数 还是 只能docker run的时候使用
> 
> 目前可以使用 tini 插件解决 就是感觉 不想安装
> 
> ```
> FROM openjdk:8-jdk-alpine3.9
> 
> COPY *.jar /app.jar
> 
> RUN apk --no-cache add tini
> 
> CMD java -Dfile.encoding=UTF-8 -Duser.timezone=PRC -Dserver.port=8080 -jar /app.jar
> 
> ENTRYPOINT ["tini"]
> ```

直接在docker run时候添加参数就行，不需要安装其他插件

**Comment by cyao2q** [comment]
> > > 
> > 
> > 
> > Dockerfile中如何使用 --init 参数 还是 只能docker run的时候使用
> > 目前可以使用 tini 插件解决 就是感觉 不想安装
> > ```
> > FROM openjdk:8-jdk-alpine3.9
> > 
> > COPY *.jar /app.jar
> > 
> > RUN apk --no-cache add tini
> > 
> > CMD java -Dfile.encoding=UTF-8 -Duser.timezone=PRC -Dserver.port=8080 -jar /app.jar
> > 
> > ENTRYPOINT ["tini"]
> > ```
> 
> 直接在docker run时候添加参数就行，不需要安装其他插件

现在是在K8S运行 那样可能需要运维修改了
